proposals/massive_context_refactor.cpp

```diff
@@ -0,0 +1,381 @@
+        operator int16_t&  ( )       { return static_cast<int16_t&> ( *reinterpret_cast<int16_t*>  ( p)); }
+
+        operator uint32_t  ( ) const { return static_cast<uint32_t> ( *reinterpret_cast<uint32_t*> ( p)); }
+        operator uint32_t& ( )       { return static_cast<uint32_t&> ( *reinterpret_cast<uint32_t*> ( p)); }
+        operator int32_t   ( ) const { return static_cast<int32_t> ( *reinterpret_cast<int32_t*>  ( p)); }
+        operator int32_t&  ( )       { return static_cast<int32_t&> ( *reinterpret_cast<int32_t*>  ( p)); }
+
+        operator float   ( ) const { return static_cast<float> ( *reinterpret_cast<float*>  ( p)); }
+        operator float&  ( )       { return static_cast<float&> ( *reinterpret_cast<float*>  ( p)); }
+
+};
+
+// template<typename Allocator=utensor::DefaultTensorMetaDataAllocator>
+class TensorBase {
+public:
+    TensorBase(){
+        utensor::Context::get_default_context().register(*this);
+    }
+    //virtual ~TensorBase(){}
+
+    // Allocate the tensor metadata on a different heap from the data scratch pads
+
+    void* operator new(size_t sz) { 
+        void* p = utensor::Context::DefaultTensorMetaDataAllocator::allocate(sz); 
+        return p;
+    }
+    void operator delete(void* p) {
+        utensor::Context::DefaultTensorMetaDataAllocator::deallocate(p);
+    }
+};
+
+class TensorInterface : public TensorBase {
+    protected:
+        virtual void* read(uint32_t linear_index) const = 0; // Handle to the data
+        virtual void* write(uint32_t linear_index) = 0
+    public:
+        ttype get_type() const { return _type; }
+        TensorShape& get_shape() { return _shape; }
+        TensorInterface() : TensorBase(), _shape(0), _type(undefined) {}
+        TensorInterface(TensorShape _shape, ttype _type) : TensorBase(), _shape(_shape), _type(_type) {}
+        virtual ~TensorInterface() {};
+
+        // Can access Tensors like
+        // mTensor(1) = 5, mTensor(2,2) = 5, etc.
+        const IntegralValue operator()(uint16_t i, uint16_t j = 0, uint16_t k = 0, uint16_t l = 0){
+            // Add shape checks here
+            return read(_shape.linear_index(i, j, k, l));
+        }
+        IntegralValue& operator()(uint16_t i, uint16_t j = 0, uint16_t k = 0, uint16_t l = 0){
+            // Add shape checks here
+            return write(_shape.linear_index(i,j,k,l));
+        }
+        virtual void resize(TensorShape new_shape) = 0;
+
+    private:
+        TensorShape _shape;
+        ttype _type; // Maybe make this const
+};
+
+class utensor::AllocatorInterface {
+
+    protected:
+        virtual void _bind(void* ptr, utensor::Tensor* hndl) = 0;
+        virtual void _unbind(void* ptr, utensor::Tensor* hndl) = 0;
+        virtual bool _is_bound(void* ptr, utensor::Tensor* hndl) = 0;
+        virtual bool _has_handle(utensor::Tensor* hndl) = 0;
+
+    public:
+        void update_hndl(Tensor& h, utensor::Tensor* new_t_ptr) {
```

Comment:
`Tensor` and `utensor::Tensor` are the same?

Reply:
I think that is the same.

Reply:
One of these is supposed to be different, let me check